pkg/controller/ingress-shim/controller_test.go

```diff
@@ -0,0 +1,170 @@
+					Namespace: "namespace-1", Name: "ingress-2",
+				}}, ingressGVK)},
+			}}},
+			givenCall: func(t *testing.T, c cmclient.Interface, _ kclient.Interface) {
+				_, err := c.CertmanagerV1().Certificates("namespace-1").Update(context.Background(), &cmapi.Certificate{ObjectMeta: metav1.ObjectMeta{
+					Namespace: "namespace-1", Name: "cert-1",
+					OwnerReferences: []metav1.OwnerReference{*metav1.NewControllerRef(&networkingv1beta1.Ingress{ObjectMeta: metav1.ObjectMeta{
+						Namespace: "namespace-1", Name: "ingress-2",
+					}}, ingressGVK)},
+				}}, metav1.UpdateOptions{})
+				require.NoError(t, err)
+			},
+			expectRequeueKey: "namespace-1/ingress-2",
+		},
+		{
+			name: "ingress is re-queued when a 'Deleted' event is received for its child Certificate",
+			existingCMObjects: []runtime.Object{&cmapi.Certificate{ObjectMeta: metav1.ObjectMeta{
+				Namespace: "namespace-1", Name: "cert-1",
+				OwnerReferences: []metav1.OwnerReference{*metav1.NewControllerRef(&networkingv1beta1.Ingress{ObjectMeta: metav1.ObjectMeta{
+					Namespace: "namespace-1", Name: "ingress-2",
+				}}, ingressGVK)},
+			}}},
+			givenCall: func(t *testing.T, c cmclient.Interface, _ kclient.Interface) {
+				err := c.CertmanagerV1().Certificates("namespace-1").Delete(context.Background(), "cert-1", metav1.DeleteOptions{})
+				require.NoError(t, err)
+			},
+			expectRequeueKey: "namespace-1/ingress-2",
+		},
+	}
+
+	for _, test := range tests {
+		t.Run(test.name, func(t *testing.T) {
+			b := &testpkg.Builder{T: t, CertManagerObjects: test.existingCMObjects, KubeObjects: test.existingKObjects}
+			b.Init()
+
+			// We don't care about the HasSynced functions since we already know
+			// whether they have been properly "used": if no Ingress or
+			// Certificate event is received then HasSynced has not been setup
+			// properly.
+			queue, _, err := (&controller{}).Register(b.Context)
+			require.NoError(t, err)
+
+			b.Start()
+			defer b.Stop()
+
+			test.givenCall(t, b.CMClient, b.Client)
+
+			// We have no way of knowing when the informers will be done adding
+			// items to the queue due to the "shared informer" architecture:
+			// Start(stop) does not allow you to wait for the informers to be
+			// done. To work around that, we do a second queue.Get and expect it
+			// to be nil.
+			time.AfterFunc(50*time.Millisecond, queue.ShutDown)
+
+			var gotKeys []string
+			for {
+				// Get blocks until either (1) a key is returned, or (2) the
+				// queue is shut down.
+				gotKey, done := queue.Get()
+				if done {
+					break
+				}
+				gotKeys = append(gotKeys, gotKey.(string))
+			}
+			assert.Equal(t, 0, queue.Len(), "queue should be empty")
+
+			// We only expect 0 or 1 keys received in the queue.
+			if test.expectRequeueKey != "" {
+				assert.Equal(t, []string{test.expectRequeueKey}, gotKeys)
+			} else {
+				assert.Nil(t, gotKeys)
+			}
```

Comment:
@irbekrm I simplified a bit with the above `for` loop that makes it clear that we are waiting until the queue is shut down by the timer.